core/state_processor.go

```diff
@@ -86,11 +105,56 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // ApplyTransactions returns the generated receipts and vm logs during the
 // execution of the state transition phase.
 func ApplyTransaction(config *ChainConfig, bc *BlockChain, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *big.Int, cfg vm.Config) (*types.Receipt, vm.Logs, *big.Int, error) {
-	_, gas, err := ApplyMessage(NewEnv(statedb, config, bc, tx, header, cfg), tx, gp)
+	env := NewEnv(statedb, config, bc, tx, header, cfg)
+	_, gas, err := ApplyMessage(env, tx, gp)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
+	// Check whether the DAO needs to be blocked or not
+	if bc != nil { // Test chain maker uses nil to construct the potential chain
+		blockRuptureCodes := false
+
+		if number := header.Number.Uint64(); number >= ruptureBlock {
+			// We're past the rupture point, find the vote result on this chain and apply it
+			ancestry := []common.Hash{header.Hash(), header.ParentHash}
+			for _, ok := ruptureGasCache[ancestry[len(ancestry)-1]]; !ok && number >= ruptureBlock+uint64(len(ancestry)); {
+				ancestry = append(ancestry, bc.GetHeaderByHash(ancestry[len(ancestry)-1]).ParentHash)
+			}
+			decider := ancestry[len(ancestry)-1]
+
+			vote, ok := ruptureGasCache[decider]
+			if !ok {
+				// We've reached the rupture point, retrieve the vote
+				vote = bc.GetHeaderByHash(decider).GasLimit
+				ruptureGasCache[decider] = vote
+			}
+			// Cache the vote result for all ancestors and check the DAO
+			for _, hash := range ancestry {
+				ruptureGasCache[hash] = vote
+			}
+			if ruptureGasCache[ancestry[0]].Cmp(ruptureThreshold) <= 0 {
+				blockRuptureCodes = true
+			}
+			// Make sure we don't OOM long run due to too many votes caching up
+			for len(ruptureGasCache) > ruptureCacheLimit {
+				for hash, _ := range ruptureGasCache {
+					delete(ruptureGasCache, hash)
+					break
+				}
+			}
+		}
+		// Verify if the DAO soft fork kicks in
+		if blockRuptureCodes {
+			if recipient := tx.To(); recipient == nil || !ruptureWhitelist[*recipient] {
```

Comment:
Just to. These are not plain addresses, rather multisig contracts. So to whitelist interaction with these contracts, the transaction `to` fields needs to be actually checked.